### ISSUE TITLE
Is it possible to support GitHub Copilot Chat?

### ISSUE THREAD

**Issue by pidgeon777** [SilasMarvin/lsp-ai#84]
As the titles says.

References:

[https://docs.github.com/en/copilot/quickstart](https://docs.github.com/en/copilot/quickstart)

[https://github.com/CopilotC-Nvim/CopilotChat.nvim](https://github.com/CopilotC-Nvim/CopilotChat.nvim)

**Comment by pidgeon777** [comment]
It is important to note that **GitHub Copilot Chat** supports various models, including `GPT-4`, `GPT-4o`, `o1`, and `o1-mini`. For more detailed information and updates, you can refer to the [GitHub Copilot Chat changelog](https://github.blog/changelog/label/copilot-chat/).

Given the widespread use and active subscriptions to **GitHub Copilot Chat** among developers, it would be highly beneficial for the `lsp-ai` plugin to incorporate support for these models. This integration would enhance the development experience by leveraging the advanced capabilities of **GitHub Copilot Chat**.

**Comment by arunoruto** [comment]
As much as I remember, Copilot is based on OpenAIs ChatGPT. I remember [`helix-gpt`](https://github.com/leona/helix-gpt) having a [GitHub Copilot](https://github.com/leona/helix-gpt/blob/master/src/providers/github.ts) integration. I looked a bit more at the code and found out, that they are using the https://api.githubcopilot.com/chat/completions endpoint. I guess you can just use the `open_ai` examples and replace the endpoint with the one from githubcopilot. [Here](https://bootk.id/posts/copilot/) is some more information about the API:

> After doing more analysis, I determined that the API endpoint at `https://api.githubcopilot.com/chat/completions` was nothing more than an OpenAI API proxy, with a different auth wrapper (which authenticated using the Copilot bearer token instead of an OpenAI key). I could query any model, with nearly identical behavior to the real OpenAI API.

I tried it out with helix, but I am getting the following error:
```log
2024-11-26T21:55:58.707 helix_lsp::transport [ERROR] lsp-ai err <- "ERROR lsp_ai::transformer_worker: generating response: error decoding response body: expected value at line 1 column 1\n"
2024-11-26T21:55:58.707 helix_lsp::transport [ERROR] lsp-ai err <- "\n"
2024-11-26T21:55:58.707 helix_lsp::transport [ERROR] lsp-ai err <- "Caused by:\n"
2024-11-26T21:55:58.707 helix_lsp::transport [ERROR] lsp-ai err <- "    expected value at line 1 column 1\n"
2024-11-26T21:55:58.707 helix_lsp::transport [ERROR] lsp-ai <- InternalError: error decoding response body: expected value at line 1 column 1

```

**EDIT** My current lsp-ai config for helix can be found in [flake repo](https://github.com/arunoruto/flake/blob/main/modules/home-manager/server/cli/helix/languages/lsp-ai.nix). You can kinda read like json, just replace the `=` with `:` and remove the `;` at the end;

**EDIT 2** I looked a bit more into the whole Open AI API compatibility. While `api.githubcopilot.com` is being used now, it used to be a different endpoint until the beginning of 2024. According to [GitHubs changelog](https://github.blog/changelog/2024-01-18-copilot-january-18th-update/), they used to have copilot-proxy.githubusercontent.com as their API endpoint. This is also referenced in [this line](https://github.com/leona/helix-gpt/blob/2a047347968e63ca55e2bf0db74e80130941f1f8/src/providers/github.ts#L168) of helix-gpt.
What does this mean? Chat should work with the https://api.githubcopilot.com/chat/completions endpoint, but completion will not. Doing some manual curl commands, the /chat/completions endpoint expects a `messages` entry in the data, while for completions a `prompt` is used. I tried manually submitting a request to https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex/completions as described by helix-gpt, but my token isn't the right one for such requests... Looking at the code, the copilot API token is used to obtain a session token, which is then used further to make requests.
tldr;
- Copilot chat should work using https://api.githubcopilot.com/chat/completions as an endpoint and the output of `gh auth token` as the token value. I used `set -gx GH_AUTH_TOKEN $(gh auth token)` to set the env variable in fish.
- Completions need a different endpoint which requires the `prompt` property to be set.

**Comment by arunoruto** [comment]
I think I got chat working using https://api.githubcopilot.com/chat/completions! The key for it to work is to set the model to be an empty string `"model": ""`. I am getting an output, even tho it is not that fast.

As for the code completion, this is gonna be tricky...
Github deviates from the OpenAI API here, and requires a new token every x minutes/hours (not sure, but it changes with time). `helix-gpt` sends a request to https://api.github.com/copilot_internal/v2/token with the copilot token (different than the one obtained with `gh auth token`) and in that response, a new token is obtained which can be used to make new requests.

tldr; chat can be used now (please test it and give feedback) and completions cannot be configured with the current state of the code (I guess?)

**EDIT**: it is possible to use the chat API to make completions as shown in the wiki: https://github.com/SilasMarvin/lsp-ai/wiki/Configuration#instruction
If someone has some nice prompts for this, enlighten us!
